### ISSUE TITLE
Project Roadmap

### ISSUE THREAD

**Issue by aikramer2** [oracle/Skater#283]
Hey all, I've had to step away from this project for some time, but I'd love to step back into it and start contributing to this project. I thought I'd begin by enumerating what I think are some key things to work on:

*Stability and bug fixes*
There are a number of bugs related to the dual support of pandas and numpy as training data arrays. I definitely think this is key draw of the library, but there are a few bugs to tackle here.

*Project management and goals*
I'd like us to maintain a clear roadmap of the project, reevaluate our contributing/review process, etc. There were a couple recent PR to the library that probably would not have been merged had they underwent more rigorous review. I also think we need to constrain the library to be as model agnostic as possible. We should offer some tooling specific to text and images, but I don't believe we should offer algorithms are specifically designed to help explain deep learning models. Given a problem domain, the goal of this library was always to provide a uniform way of explaining model behavior while being totally agnostic about what the model is actually doing under the hood. Providing tooling specific to neural networks goes against that idea. Another aspect to review / potentially formalize is committing to "global" interpretation algorithms only. We initially included lime to help make skater a "one stop shop" for model interpretation and to have some kind of local interpretation offering, but our support for LIME has always been janky, and its added an extra layer of work trying to keep our library current with updates to the lime package. This was a mistake, and we ought to simply reference and support their package as another helpful model interpretation library with slightly different goals.

* Testing suite *
Ya, more tests :)

* Visualization Improvements*
There are a few issues with our visualization suite that are worth reviewing: 
- default scientific notation
- auto-scaling ylims (which can be very misleading on pdp plots).
- alternative options for bivariate PDP plots (heatmap style in addition to the current 3D offering)

* Better documentation related to input_formatters*
One of the most powerful abstractions in skater is that you can run perturbations over an input prior to passing data through a preprocessing pipeline. A typical use case would be that you use an input formatter to one hot encode your data such that you can estimate the feature importance of all of the encoded values as a single variable, even if your model treats them independently. Its not clear to me that people understand this workflow, and perhaps we could upgrade our tooling and documentation to make this more obvious.

* Algorithms *
One key limitation to the current workflow with skater is that the user doesn't know which combinations of features are worth exploring via bivariate pdp plots. We could close the loop here by offering some kind of interaction detection algorithms. There are a few families of approaches here, I'd be happy to start a project related to interaction detection support.

Would love to get some feedback from the current admins of this repo, as the library could use a lot of attention, and its now falling out of use:

https://github.com/christophM/interpretable-ml-book/issues/124

**Comment by darenr** [comment]
Thanks Aaron, always willing to merge your PRs. The model-specific merge was thought to be a useful addition (as would other model-specific approaches) because I don't believe there's anything special about the idea of always treating a model as a black box, there are times when more information is available and aggregating over some sample of the training data to arrive at a global explanation will never be as fast or accurate as taking advantage of the model-specific knowledge often available. For DL CAMs are particularly useful for example. However, your roadmap looks great, couldn't agree more about very misleading pdp plots (pdp plots are already very misleading to start with without adding to the problem).

I think the two biggest issues with Skater are: 1. LIME dependency (which you've addressed) and 2. keeping up to date with other changes in dependent packages specifically for visualizations where there's a lot of change happening.

**Comment by aikramer2** [comment]
Thanks @darenr. Re: roadmap, i'd love to get a doc going and start assigning the tasks.

I do disagree that there's nothing special about treating a model as a blackbox, and that that shouldn't be a constraint of the library. The the main Skater's API ( all algorithms in the interpreter object supporting anything that implements the model and dataset interface) was intended to target model agnostic algorithms (its referenced in the abstract of the package :) ). What is special about treating models as black boxes is that any algorithm we develop today is guaranteed to work on all models, including those that havent been invented yet, and that it allows you to create tooling that has uniform API's across models. I totally agree CAMs are useful, but I dont think that they fit well in this library, either conceptually or in the design of the software. Further, we're we to commit ourselves to algorithm specific interpretations algorithms and visualizations, we'd either: 

- explode the scope of the project start with model specific interpretation graphs and algorithms.
or
- make subjective choices around which model specific algorithms and visualizations to support (probably based off surveying whats available in the ecosystem and reviewing).

Both of these options would be difficult to support well given the lack of strong contributor base. Let me know what you think.

**Comment by darenr** [comment]
We can keep Skater to model agnostic, however, the approach of global interpretation as a function of the aggregate of local using model agnostic approaches can often be misleading. But given the lack of strong contributor base, staying focused on keeping things small is worthwhile.

Getting started with a roadmap doc would be a good idea. Oracle will also help as much as the team here can. One place I'd suggest to improve is to offer (or switch to) ALE plots which are a faster and are an unbiased alternative to partial dependence plots. There was only an R implementation but now there's also [ALEPython](https://github.com/blent-ai/ALEPython) which might be worth taking a look at and using.

For this to continue to be an Oracle offering we should make sure that it's kept up to date with python, dropping 2.7 at some point and ensuring all tests run on 3.X, currently it's 2.7, 3.5, 3.6 but aligning with LIME's 3.6, 3.7 only would make sense. I've also been asked about a docker container with everything packaged, something else to consider.
